src/utilities/xml/Validator.cpp

```diff
@@ -0,0 +1,143 @@
+*  ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***********************************************************************************************************************/
+
+#include "Validator.hpp"
+
+#include <xercesc/util/XMLString.hpp>
+#include <xercesc/parsers/XercesDOMParser.hpp>
+#include <xercesc/framework/LocalFileInputSource.hpp>
+#include <xercesc/sax/ErrorHandler.hpp>
+#include <xercesc/sax/SAXParseException.hpp>
+#include <xercesc/validators/common/Grammar.hpp>
+
+namespace openstudio {
+
+class ParserErrorHandler : public xercesc::ErrorHandler
+{
+ private:
+  void reportParseException(const xercesc::SAXParseException& ex) {
+    char* msg = xercesc::XMLString::transcode(ex.getMessage());
+    fprintf(stderr, "at line %llu column %llu, %s\n", ex.getLineNumber(), ex.getColumnNumber(), msg);
+    xercesc::XMLString::release(&msg);
+  }
+
+ public:
+  void warning(const xercesc::SAXParseException& ex) {
+    reportParseException(ex);
+  }
+
+  void error(const xercesc::SAXParseException& ex) {
+    reportParseException(ex);
+  }
+
+  void fatalError(const xercesc::SAXParseException& ex) {
+    reportParseException(ex);
+  }
+
+  void resetErrors() {}
+};
+
+Validator::Validator(const openstudio::path& xsdPath) : m_xsdPath(openstudio::filesystem::system_complete(xsdPath)) {
+  if (!openstudio::filesystem::exists(xsdPath)) {
+    LOG_AND_THROW("'" << toString(xsdPath) << "' does not exist");
+  } else if (!openstudio::filesystem::is_regular_file(xsdPath)) {
+    LOG_AND_THROW("'" << toString(xsdPath) << "' XSD cannot be opened");
+  }
+
+  xercesc::XMLPlatformUtils::Initialize();
+
+  // create the parser here?
+}
+
+Validator::Validator(const std::string& xsdString) : m_xsdString(xsdString) {
+  // create the parser here?
+}
+
+openstudio::path Validator::xsdPath() const {
+  return m_xsdPath;
+}
+
+std::vector<std::string> Validator::errors() const {
+  return m_errors;
+}
+
+std::vector<std::string> Validator::warnings() const {
+  return m_warnings;
+}
+
+bool Validator::isValid() const {
+  return (m_errors.size() == 0u);
+}
+
+bool Validator::validate(const openstudio::path& xmlPath) const {
+  if (!openstudio::filesystem::exists(xmlPath)) {
+    LOG(Error, "'" << toString(xmlPath) << "' does not exist");
+    return false;
+  } else if (!openstudio::filesystem::is_regular_file(xmlPath)) {
+    LOG(Error, "'" << toString(xmlPath) << "' XML cannot be opened");
+    return false;
+  }
+
+  xercesc::XercesDOMParser parser;
+
+  // Let's preparse the schema grammar (.xsd) and cache it.
+  if (parser.loadGrammar(toString(m_xsdPath).c_str(), xercesc::Grammar::SchemaGrammarType) == NULL) {
+    LOG(Error, "'" << toString(m_xsdPath) << "' XSD cannot be loaded");
+    return false;
+  }
+
+  // Enable schema processing.
+  parser.setDoSchema(true);
+  parser.setDoNamespaces(true);
+  parser.setValidationConstraintFatal(true);
+  parser.setValidationSchemaFullChecking(true);
```

Comment:
Do we need `parser.setValidationScheme(XercesDOMParser::Val_Always);` ?

https://xerces.apache.org/xerces-c/apiDocs-3/classAbstractDOMParser.html#a35600d583c42ceea9b04ee10814238a6

https://xerces.apache.org/xerces-c/apiDocs-3/classAbstractDOMParser.html#a76513b711425077116569013c93a943c